src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+    (0001-01-01) in dates.
+]]
+
+ffi.cdef [[
+
+/* dt_core.h definitions */
+typedef int dt_t;
+
+typedef enum {
+    DT_MON       = 1,
+    DT_MONDAY    = 1,
+    DT_TUE       = 2,
+    DT_TUESDAY   = 2,
+    DT_WED       = 3,
+    DT_WEDNESDAY = 3,
+    DT_THU       = 4,
+    DT_THURSDAY  = 4,
+    DT_FRI       = 5,
+    DT_FRIDAY    = 5,
+    DT_SAT       = 6,
+    DT_SATURDAY  = 6,
+    DT_SUN       = 7,
+    DT_SUNDAY    = 7,
+} dt_dow_t;
+
+dt_t   tnt_dt_from_rdn     (int n);
+dt_t   tnt_dt_from_ymd     (int y, int m, int d);
+void   tnt_dt_to_ymd       (dt_t dt, int *y, int *m, int *d);
+void   tnt_dt_to_yqd       (dt_t dt, int *y, int *q, int *d);
+void   tnt_dt_to_ywd       (dt_t dt, int *y, int *w, int *d);
+
+int    tnt_dt_rdn          (dt_t dt);
+dt_dow_t tnt_dt_dow        (dt_t dt);
+
+/* dt_util.h */
+bool    tnt_dt_leap_year       (int y);
+int     tnt_dt_days_in_year    (int y);
+int     tnt_dt_days_in_quarter (int y, int q);
+int     tnt_dt_days_in_month   (int y, int m);
+int     tnt_dt_weeks_in_year   (int y);
+
+/* dt_accessor.h */
+
+int     tnt_dt_year         (dt_t dt);
+int     tnt_dt_month        (dt_t dt);
+int     tnt_dt_doy          (dt_t dt);
+int     tnt_dt_dom          (dt_t dt);
+
+/* dt_arithmetic.h definitions */
+
+typedef enum {
+    DT_EXCESS,
+    DT_LIMIT,
+    DT_SNAP
+} dt_adjust_t;
+
+dt_t   tnt_dt_add_years    (dt_t dt, int delta, dt_adjust_t adjust);
+dt_t   tnt_dt_add_quarters (dt_t dt, int delta, dt_adjust_t adjust);
+dt_t   tnt_dt_add_months   (dt_t dt, int delta, dt_adjust_t adjust);
+
+/* dt_parse_iso.h definitions */
+
+size_t tnt_dt_parse_iso_date (const char *str, size_t len, dt_t *dt);
+size_t tnt_dt_parse_iso_time (const char *str, size_t len, int *sod, int *nsec);
+size_t tnt_dt_parse_iso_zone_lenient(const char *str, size_t len, int *offset);
+
+/* Tarantool functions - datetime.c */
+
+int    datetime_to_string(char *buf, int len, const struct datetime * date);
+size_t datetime_strftime(char *buf, uint32_t len, const char *fmt,
+                        const struct datetime *date);
+void   datetime_now(struct datetime *now);
+
+]]
+
+local builtin = ffi.C
+local math_modf = math.modf
+local math_floor = math.floor
+
+local SECS_PER_DAY     = 86400
+local NANOS_PER_SEC    = 1000000000
+
+-- c-dt/dt_config.h
+
+-- Unix, January 1, 1970, Thursday
+local DT_EPOCH_1970_OFFSET = 719163
+
+
+local datetime_t = ffi.typeof('struct datetime')
+
+local function is_datetime(o)
+    return ffi.istype(datetime_t, o)
+end
+
+local function check_date(o, message)
+    if not is_datetime(o) then
+        return error(("%s: expected datetime, but received %s"):
+                     format(message, o), 2)
+    end
+end
+
+local function check_table(o, message)
+    if type(o) ~= 'table' then
+        return error(("%s: expected table %s"):format(message, o), 2)
+    end
+end
+
+local function check_str(s, message)
+    if not type(s) == 'string' then
```

Comment:
`not type() ==` is not the same as `type() ~=`. Please, use the second one. And cover it with a test. If you left it like that, it means there is no a test. For example, it should raise an error for a number, but it won't:
```
> not 'number' == 'string'
false
> 'number' ~= 'string'
true
```

Reply:
Thanks! Corrected. Actually that code was called from parse_zone according to `luacov` report I see:

```
     local function check_str(s, message)
   8     if type(s) ~= 'string' then
***0         return error(("%s: expected string, but received %s"):
***0                      format(message, s), 2)
         end
     end
    ...
     local parse_zone = function(str)
   8     check_str("datetime.parse_zone()")
   8     local offset = ffi.new('int[1]')
   8     local len = builtin.tnt_dt_parse_iso_zone_lenient(str, #str, offset)
   8     if len == 0 then
   6         error(('invalid time-zone format %s'):format(str), 3)
         end
   2     return offset[0]
     end

```

For this commit coverage is 
```
builtin/datetime.lua 203  57     78.08%
```
within full branch code it is increased to 84%

Reply:
The report above says the error is never raised: `***0`. Please, cover it with a test.

Reply:
Covered